### PR TITLE
Outsource db connection along with models

### PR DESCRIPTION
**Pull request by HuangStanley050**
* All DB connections will be made using connectToDatabase()
* Made/outsourced User and Hack models for easier DB manipulation

### affected lambdas are:

1. userHacks
2. authUser
3. registerUser
4. getHacks
5. hackDetails
6. editHack
7. joinHack
8. addHack

### CODE REVIEW

File: services/database/db.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  if (isConnected) {
+    console.log("=> using existing database connection");
+    return Promise.resolve();
+  }
+
+  console.log("=> using new database connection");
+  return mongoose.connect(url).then(db => {
+    isConnected = db.connections[0].readyState;
+  });
+};
```
Comment (jpeyper) [inline services/database/db.js]: could be

```
export const connectToDatabase = async () => {
  if (isConnected) {
    console.log("=> using existing database connection");
    return;
  }

  console.log("=> using new database connection");
  const db = await mongoose.connect(url);
  isConnected = db.connections[0].readyState;
  return
};
```

Comment (HuangStanley050) [reply services/database/db.js]: Thank you for showing me how to convert a promise function to async and await!

File: services/database/models/HackModel.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import mongoose from "mongoose";
+
+const Schema = mongoose.Schema;
+const hackSchema = new Schema({
+  title: { type: String, default: "" },
+  description: { type: String, default: "" },
+  goal: { type: String, default: "" },
+  team: { type: [Schema.Types.ObjectId], default: [] },
```
Comment (jpeyper) [inline services/database/models/HackModel.js:8]: should this be defined as

```
team: { type: [Schema.Types.ObjectId], ref: 'User', default: [] },
```

(and the other two)

https://mongoosejs.com/docs/populate.html

Comment (HuangStanley050) [reply services/database/models/HackModel.js:8]: That's what I originally thought as well, appears the team have changed their approach, that's why I used populate below to refer to the model:

```javascript
  let result = await Hack.findById(id)
      .select("-likes")
      .populate("team", "-email", User);
```

https://stackoverflow.com/questions/14627948/mongoose-populate-field-without-ref-option

File: services/database/db.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import mongoose from "mongoose";
+mongoose.Promise = global.Promise;
+
+let isConnected;
+const url = `mongodb://${process.env.MONGO_USER}:${process.env.MONGO_PASSWORD}@ds157136.mlab.com:57136/hackone`;
+
+export const connectToDatabase = () => {
+  if (isConnected) {
```
Comment (jpeyper) [inline services/database/db.js:8]: lets not use a stored value that is not kept in sync:

https://mongoosejs.com/docs/api/mongoose.html#mongoose_Mongoose-connection

```
if (mongoose.connection.readyState === 1)
```

Although I'm wondering how much we need to do this... can we call it once during the execution of our function (outside the function, not inside) (`keepAlive` is on by default: https://mongoosejs.com/docs/connections.html#keepAlive)

The commands have buffering enabled by default (https://mongoosejs.com/docs/connections.html#buffering) so we dont need to wait for it to connect before moving on here.

**Review by jpeyper — APPROVED** [review]

File: services/database/db.js (hunk 3 above)
Comment (HuangStanley050) [reply services/database/db.js:8]: Not able to find an example online that calls the connection outside of the lambda function, the official recommendation from Mongo seems to also advocate using connection inside the function. At the moment it seems okay.

https://docs.atlas.mongodb.com/best-practices-connecting-to-aws-lambda/

Comment (jpeyper) [reply services/database/db.js:8]: interesting find :)

**Pull request merged**